File: benchmarks/database-core-queries.ts
/** Cross-binding read benchmark — TypeScript lane. */
import { mkdirSync, writeFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'
import { PolyGraph } from '../src/graph.js'
import { VectorIndex } from '../src/vector-index.js'
import { HNSWIndex } from '../src/hnsw-index.js'
import { MemoryAdapter } from '../src/persistence/memory.js'

const dir = dirname(fileURLToPath(import.meta.url))
const arg = (name: string, fallback: number) => { const i = process.argv.indexOf(name); return i >= 0 ? Number(process.argv[i + 1]) : fallback }
const count = arg('--count', 10_000)
const iterations = arg('--iterations', 20)
const dims = 32
const topK = 10
const dataSeed = 42
const querySeed = 43

function mulberry32(seed: number): () => number { let state = seed >>> 0; return () => { state = (state + 0x6d2b79f5) | 0; let t = Math.imul(state ^ (state >>> 15), 1 | state); t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t; return ((t ^ (t >>> 14)) >>> 0) / 4294967296 } }
function vectors(count: number, seed: number): number[][] { const rng = mulberry32(seed); return Array.from({ length: count }, () => Array.from({ length: dims }, () => rng() * 2 - 1)) }
function node(i: number, values: number[]) { return { id: `n${i}`, type: ['user', 'post', 'comment'][i % 3], data: { score: i % 1000, bucket: i % 10, value: i }, vector: values, insertedAt: i, updatedAt: i } }
function stats(times: number[]) { const s = [...times].sort((a, b) => a - b); const at = (p: number) => s[Math.min(s.length - 1, Math.ceil(s.length * p) - 1)]; return { p50Ms: at(.5), p95Ms: at(.95), p99Ms: at(.99) } }
function measure(fn: () => unknown) { const times: number[] = []; let value: any; for (let i = 0; i < iterations; i++) { const t = performance.now(); value = fn(); times.push(performance.now() - t) } return { ...stats(times), value } }

async function main() {
  if (!Number.isInteger(count) || count < 1) throw new Error('--count must be positive')
  const outArg = process.argv.indexOf('--out')
  const out = outArg >= 0 ? process.argv[outArg + 1] : join(dir, 'results/database-core-queries-ts.json')
  const data = vectors(count, dataSeed)
  const graph = new PolyGraph(new MemoryAdapter())
  for (let i = 0; i < count; i++) graph.addNode(node(i, data[i]))
  await graph.flush()
  const expected = count <= 22 ? 0 : Math.min(25, Math.floor((count - 23) / 30) + 1)
  const graphQuery = measure(() => graph.query().whereNodeType('post').where('bucket', 2).orderBy('score', 'desc').limit(25).ids())
  if (graphQuery.value.length !== expected) throw new Error(`unexpected graph query result count ${graphQuery.value.length}`)
  const hot = measure(() => graph.query().whereNodeType('post').where('bucket', 2).orderBy('score', 'desc').limit(25).ids())
  if (hot.value.length !== graphQuery.value.length) throw new Error('graph query result mismatch')

  const query = vectors(1, querySeed)[0]
  const exact = new VectorIndex(); exact.addMany(data.map((value, i) => ({ id: `n${i}`, vector: value })))
  const exactMeasured = measure(() => exact.query(query, topK))
  if (exactMeasured.value.length !== topK) throw new Error('exact vector result verification failed')
  const hnsw = new HNSWIndex(undefined, undefined, { efSearch: 300 }, mulberry32(7))
  hnsw.addMany(data.map((value, i) => ({ id: `n${i}`, vector: value })))
  const hnswMeasured = measure(() => hnsw.query(query, topK))
  if (hnswMeasured.value.length !== topK) throw new Error('HNSW result verification failed')
  const exactIds = new Set(exactMeasured.value.map((x: any) => x.id)); const recall = hnswMeasured.value.filter((x: any) => exactIds.has(x.id)).length / topK
  const result = { schemaVersion: 1, lang: 'ts', count, dimensions: dims, iterations, topK, dataSeed, querySeed, graphQuery: { ...graphQuery, value: undefined, resultCount: graphQuery.value.length }, hotQuery: { ...hot, value: undefined, resultCount: hot.value.length }, exactVector: { ...exactMeasured, value: undefined, resultCount: topK }, hnswVector: { ...hnswMeasured, value: undefined, resultCount: topK, recallAtK: recall } }
  mkdirSync(dirname(out), { recursive: true }); writeFileSync(out, JSON.stringify(result, null, 2)); console.log(JSON.stringify(result, null, 2)); console.log(`Wrote ${out}`)
}
main().catch(error => { console.error(error); process.exit(1) })

File: crates/polypack-graph/examples/database_core_queries.rs
//! Cross-binding read benchmark — Rust lane.

use polypack_core::{DistanceFn, ExactIndex, HnswConfig, InMemoryStorage, Node, StoreConfig};
use polypack_graph::{Graph, GraphConfig, OrderDirection};
use serde_json::json;
use std::env;
use std::fs;
use std::path::PathBuf;
use std::time::Instant;

const DIMS: usize = 32;
const TOP_K: usize = 10;

struct Args { count: usize, iterations: usize, out: PathBuf }
fn args() -> Args {
    let raw: Vec<String> = env::args().collect(); let mut count = 10_000; let mut iterations = 20; let mut out = PathBuf::from("benchmarks/results/database-core-queries-rust.json"); let mut i = 1;
    while i < raw.len() { match raw[i].as_str() { "--count" => { count = raw[i+1].parse().unwrap(); i += 2; }, "--iterations" => { iterations = raw[i+1].parse().unwrap(); i += 2; }, "--out" => { out = raw[i+1].clone().into(); i += 2; }, _ => i += 1 } }
    Args { count, iterations, out }
}
struct Rng(u32);
impl Rng { fn next(&mut self) -> f64 { self.0 = self.0.wrapping_add(0x6d2b79f5); let mut t = (self.0 ^ (self.0 >> 15)).wrapping_mul(self.0 | 1); t = (t.wrapping_add((t ^ (t >> 7)).wrapping_mul(self.0 | 61))) ^ t; ((t ^ (t >> 14)) as f64) / 4294967296.0 } }
fn vectors(count: usize, seed: u32) -> Vec<Vec<f64>> { let mut rng = Rng(seed); (0..count).map(|_| (0..DIMS).map(|_| rng.next() * 2.0 - 1.0).collect()).collect() }
fn node(i: usize, value: Vec<f64>) -> Node { Node { id: format!("n{i}"), node_type: ["user", "post", "comment"][i % 3].into(), data: { let mut d = serde_json::Map::new(); d.insert("score".into(), json!(i % 1000)); d.insert("bucket".into(), json!(i % 10)); d.insert("value".into(), json!(i)); d }, vector: Some(value), inserted_at: i as i64, updated_at: i as i64, revision: 0, activation: None, ..Default::default() } }
fn summary(mut times: Vec<f64>) -> serde_json::Value { times.sort_by(|a,b| a.partial_cmp(b).unwrap()); let at = |p: f64| times[((times.len() as f64 * p).ceil() as usize).saturating_sub(1).min(times.len()-1)]; json!({"p50Ms": at(0.5), "p95Ms": at(0.95), "p99Ms": at(0.99)}) }
fn main() {
    let a = args(); assert!(a.count > 0); let data = vectors(a.count, 42); let query = vectors(1, 43).remove(0); let config = GraphConfig { hnsw: HnswConfig { ef_search: 300, ..Default::default() }, ..Default::default() }; let mut graph = Graph::open(Box::new(InMemoryStorage::new()), StoreConfig::default(), config).unwrap();
    graph.add_nodes(data.iter().enumerate().map(|(i, value)| node(i, value.clone())).collect()).unwrap(); graph.flush().unwrap();
    let expected = std::cmp::min(25, if a.count <= 22 { 0 } else { (a.count - 23) / 30 + 1 });
    let mut times = Vec::new(); let mut query_count = 0;
    for _ in 0..a.iterations { let t=Instant::now(); query_count = graph.query().where_node_type(vec!["post".into()]).where_field("bucket", json!(2)).order_by("score", OrderDirection::Desc).limit(25).ids().len(); times.push(t.elapsed().as_secs_f64()*1000.0); }
    assert_eq!(query_count, expected);
    let q = query; let mut exact = ExactIndex::new(DistanceFn::Cosine); for (i, value) in data.iter().enumerate() { exact.add(&format!("n{i}"), value).unwrap(); }
    let mut exact_times=Vec::new(); let mut exact_count=0; for _ in 0..a.iterations { let t=Instant::now(); exact_count=exact.query(&q, TOP_K, 0.0).unwrap().len(); exact_times.push(t.elapsed().as_secs_f64()*1000.0); } assert_eq!(exact_count, TOP_K);
    let mut hnsw_times=Vec::new(); let mut hnsw_count=0; for _ in 0..a.iterations { let t=Instant::now(); hnsw_count=graph.similar_to(&q, 0.0, Some(TOP_K)).unwrap().len(); hnsw_times.push(t.elapsed().as_secs_f64()*1000.0); } assert_eq!(hnsw_count, TOP_K);
    let exact_ids: std::collections::HashSet<_> = exact.query(&q, TOP_K, 0.0).unwrap().into_iter().map(|x| x.id).collect(); let hnsw_ids = graph.similar_to(&q, 0.0, Some(TOP_K)).unwrap(); let recall = hnsw_ids.iter().filter(|x| exact_ids.contains(&x.0)).count() as f64 / TOP_K as f64;
    let mut graph_query = summary(times); graph_query["resultCount"] = json!(query_count);
    let mut exact_result = summary(exact_times); exact_result["resultCount"] = json!(TOP_K);
    let mut hnsw_result = summary(hnsw_times); hnsw_result["resultCount"] = json!(TOP_K); hnsw_result["recallAtK"] = json!(recall);
    let result=json!({"schemaVersion":1,"lang":"rust","count":a.count,"dimensions":DIMS,"iterations":a.iterations,"topK":TOP_K,"dataSeed":42,"querySeed":43,"graphQuery": graph_query,"hotQuery": {"resultCount":query_count},"exactVector": exact_result,"hnswVector": hnsw_result});
    if let Some(parent)=a.out.parent(){fs::create_dir_all(parent).unwrap();} fs::write(&a.out, serde_json::to_string_pretty(&result).unwrap()).unwrap(); println!("{}",serde_json::to_string_pretty(&result).unwrap()); println!("Wrote {}",a.out.display());
}
